### PR TITLE
Instrument our build so we can easily capture linker execution times

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 1]
```diff
@@ -253,9 +251,10 @@ private[sbtplugin] object ScalaJSPluginInternal {
 
             val out = linkerImpl.outputDirectory(outputDir.toPath)
 
+            val tlog = scalaJSLoggerFactory.value(log)
```
Comment (sjrd) [inline sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:254]: This should go outside the `FileFunction.cached` to please sbt (at least the `scalaJSLoggerFactory.value` part of it): `.value` calls are supposed to go outside of control-flow and closures.

File: project/Build.scala [hunk 2]
```diff
@@ -119,6 +129,38 @@ object MyScalaJSPlugin extends AutoPlugin {
...
+
+          case Some(file) =>
+            /* Instrument logger to dump calls to `time` to a file.
+             * The way we manage the file is a hack: It is difficult to properly
+             * manage resources in sbt's settings system, so we simply set the
+             * printStream to autoFlush, avoiding that we have to close the file.
+             */
+            val stream = new PrintStream(new FileOutputStream(file, /*append=*/true), /*autoFlush=*/true)
+
+            (sbtLogger: SbtLogger) => {
```
Comment (sjrd) [inline project/Build.scala:146]: Coding style nit:
```suggestion
            { (sbtLogger: SbtLogger) =>
```

File: project/Build.scala [hunk 3]
```diff
@@ -119,6 +129,38 @@ object MyScalaJSPlugin extends AutoPlugin {
...
+              val base = default(sbtLogger)
+              new Logger {
+                def log(level: Level, message: => String): Unit =
+                  base.log(level, message)
+
+                def trace(t: => Throwable): Unit =
+                  base.trace(t)
+
+                override def time(title: String, nanos: Long): Unit = {
+                  stream.println(s"$title, $nanos")
```
Comment (sjrd) [inline project/Build.scala:156]: If the output is meant to be CSV, shouldn't it be
```suggestion
                  stream.println(s"$title,$nanos")
```
instead?

**Review by sjrd — CHANGES_REQUESTED** [review]

File: project/Build.scala [hunk 4]
```diff
@@ -119,6 +129,38 @@ object MyScalaJSPlugin extends AutoPlugin {
...
+        ExposedValues.autoImport.loggerTimingOut.?.value match {
+          case None => default
+
+          case Some(file) =>
+            /* Instrument logger to dump calls to `time` to a file.
+             * The way we manage the file is a hack: It is difficult to properly
+             * manage resources in sbt's settings system, so we simply set the
+             * printStream to autoFlush, avoiding that we have to close the file.
+             */
+            val stream = new PrintStream(new FileOutputStream(file, /*append=*/true), /*autoFlush=*/true)
```
Comment (sjrd) [inline project/Build.scala:144]: Perhaps specify an explicit `StandardCharsets.UTF_8` here as well. On Windows it's typically not the default (at least before JDK 20ish).

File: project/Build.scala (hunk 2 above)
Comment (gzm0) [reply project/Build.scala:146]: I'm writing too much Typescript these days ;-)

File: project/Build.scala (hunk 3 above)
Comment (gzm0) [reply project/Build.scala]: Yes, indeed.

File: project/Build.scala (hunk 4 above)
Comment (gzm0) [reply project/Build.scala]: Good point. Needed to specify a string, since `PrintStream` doesn't accept a Charset (at least on JDK 8).

**Comment by gzm0** [comment]
I have added an additional commit that allows to add a "variant" (to a fixed file name). IMO this lets us output data in a much simpler shape for processing (e.g. when using `R`, the variant can be loaded as a factor right away, so no reshaping / merging of files is required).

**Comment by gzm0** [comment]
If you think we should keep the last commit, I'll squash before merging.

**Comment by sjrd** [comment]
The `variant` one looks good, yes.

**Comment by gzm0** [comment]
Squashed. I've also added an example of how to load and plot this data in R (with tidyverse).

**Review by sjrd — CHANGES_REQUESTED** [review]

File: sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 5]
```diff
@@ -234,6 +234,8 @@ private[sbtplugin] object ScalaJSPluginInternal {
 
         Def.task {
           val log = s.log
+          val tlog = scalaJSLoggerFactory.value(log)
```
[the remark this answers is not in the document]
Comment (gzm0) [reply sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:237]: :facepalm: fixed.

**Review by sjrd — APPROVED** [review]

**Pull request merged**